qupath-extension-openslide/src/main/java/qupath/lib/images/servers/openslide/OpenSlide.java

```diff
@@ -0,0 +1,491 @@
+        }
+
+        @Override
+        public String getDescription() {
+            return "Virtual slide";
+        }
+    };
+
+    private static final String LIBRARY_VERSION = OpenSlideJNA.INSTANCE.openslide_get_version();
+
+    final public static String PROPERTY_NAME_COMMENT = "openslide.comment";
+
+    final public static String PROPERTY_NAME_VENDOR = "openslide.vendor";
+
+    final public static String PROPERTY_NAME_QUICKHASH1 = "openslide.quickhash-1";
+
+    final public static String PROPERTY_NAME_BACKGROUND_COLOR = "openslide.background-color";
+
+    final public static String PROPERTY_NAME_OBJECTIVE_POWER = "openslide.objective-power";
+
+    final public static String PROPERTY_NAME_MPP_X = "openslide.mpp-x";
+
+    final public static String PROPERTY_NAME_MPP_Y = "openslide.mpp-y";
+
+    final public static String PROPERTY_NAME_BOUNDS_X = "openslide.bounds-x";
+
+    final public static String PROPERTY_NAME_BOUNDS_Y = "openslide.bounds-y";
+
+    final public static String PROPERTY_NAME_BOUNDS_WIDTH = "openslide.bounds-width";
+
+    final public static String PROPERTY_NAME_BOUNDS_HEIGHT = "openslide.bounds-height";
+
+    private long osr;
+
+    final private ReadWriteLock lock = new ReentrantReadWriteLock();
+
+    final private long[] levelWidths;
+
+    final private long[] levelHeights;
+
+    final private double[] levelDownsamples;
+
+    final private int levelCount;
+
+    final private Map<String, String> properties;
+
+    final private Map<String, AssociatedImage> associatedImages;
+
+    final private File canonicalFile;
+
+    final private int hashCodeVal;
+
+    public static String detectVendor(File file) {
+        return OpenSlideJNA.INSTANCE.openslide_detect_vendor(file.getPath());
+    }
+
+    public OpenSlide(File file) throws IOException {
+        if (!file.exists()) {
+            throw new FileNotFoundException(file.toString());
+        }
+
+        osr = OpenSlideJNA.INSTANCE.openslide_open(file.getPath());
+
+        if (osr == 0) {
+            throw new IOException(file
+                    + ": Not a file that OpenSlide can recognize");
+        }
+        // dispose on error, we are in the constructor
+        try {
+            checkError();
+        } catch (IOException e) {
+            dispose();
+            throw e;
+        }
+
+        // store level count
+        levelCount = OpenSlideJNA.INSTANCE.openslide_get_level_count(osr);
+
+        // store dimensions
+        levelWidths = new long[levelCount];
+        levelHeights = new long[levelCount];
+        levelDownsamples = new double[levelCount];
+
+        for (int i = 0; i < levelCount; i++) {
+            long[] w = new long[1], h = new long[1];
+            OpenSlideJNA.INSTANCE.openslide_get_level_dimensions(osr, i, w, h);
+            levelWidths[i] = w[0];
+            levelHeights[i] = h[0];
+            levelDownsamples[i] = OpenSlideJNA.INSTANCE.openslide_get_level_downsample(
```

Comment:
We *could* get the levels, but I'm not sure we should because it can give the wrong idea of what the appropriate level is for a resolution.

It's calculated by [averaging the width and height ratios](https://github.com/openslide/openslide/blame/cba7702258f777bd122432642f50d50aea0ef374/src/openslide.c#L254). However that can cause subtle problems due to rounding having forced the actual pixel dimensions to be slightly off what the (presumably) intended downsample is, and also from the width and height ratios often being different.

QuPath uses different logic that attempts to correct for this [here](https://github.com/qupath/qupath/blob/533ce0cf50ca2a5627b55a8c1fe351cdc72e0694/qupath-core/src/main/java/qupath/lib/images/servers/ImageServerMetadata.java#L1107).

Reply:
I guess we can just remove any of the openslide logic for calculating downsamples? It doesn't seem to be used.